Improved maxval, minval and sum to code transformation (closes #2323)

rupertford opened the pull request:
More featured implementation of the maxval, minval and sum to code transformations (and now a product transformation too). Adds supports for expressions within the intrinsic and mask argument and constant dimensions for any of the arrays. This should support most, if not all of the nemo examples. The main thing missing is that there is no support for the dimension argument. This implementation is a complete rework based on @sergisiso's suggestions, so if it works well I've done a great job and if not he has given bad advice.

rupertford commented:
The change to `nemo_arrayrange2loop_trans.py` and the addition of a test in the associated test file is due to a small bug that was found whilst implementing this PR. The problem was that the matching was matching the first index that had the same values as the match so if the indices had the same values we could get the wrong index returned. The change fixes this so that the correct matching index is returned.

rupertford commented:
Ready for first review. Probably one for @sergisiso seeing as this implementation is based on his suggestions.

sergisiso commented on src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py:
```diff
@@ -203,170 +187,97 @@ def apply(self, node, options=None):
-                    # reduction over so do not loop over it.
+        expr, _, mask_ref = self._get_args(node)
+
+        # Step 1 extract the expression within the intrinsic and put
```
I think each Step comment would benefit with an example comment line similar to https://github.com/stfc/PSyclone/issues/2323#issuecomment-1784897338
otherwise is a bit difficult to follow each step.

rupertford replied:
OK, I've hopefully done that to your satisfaction.

sergisiso commented on src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py:
```diff
@@ -203,170 +187,97 @@ def apply(self, node, options=None):
+        assignment_position = assignment.position
+        # Must be placed here to avoid circular imports
+        # pylint: disable=import-outside-toplevel
+        from psyclone.domain.nemo.transformations import \
+            NemoAllArrayRange2LoopTrans
```
Again, try with the whole from path, I don't see why this imports should cycle (other than the folder has an aggregation of thing that might)

rupertford replied:
There is a circular import problem here so it needs to remain as it is. The circular import makes sense I think as we are using a domain-specific transformation from a more generic PSyIR transformation and in general we would expect domain-specific transformations to import PSyIR transformations.

sergisiso commented on src/psyclone/tests/psyir/transformations/intrinsics/mms_base_trans_test.py:
```diff
@@ -445,318 +292,305 @@ def test_apply_multi(fortran_reader, fortran_writer, tmpdir):
         "  real :: value1, value2\n"
         "  real :: result\n"
-        "  result = value1 + sum(array,dim=1) * value2\n"
+        "  result = value1 + maxval(array,dim=1) * value2\n"
```
Remove the value1 and value2 from this expression to make it the simplest example of the transformationerror.

rupertford replied:
Simplified. Also removed the comment explaining the hierarchy as we use a walk.

sergisiso reviewed (changes requested):
> so if it works well I've done a great job and if not he has given bad advice.

@rupertford You have done a great job :)

I agree with your concerns regarding the naming, if by "2code" we want to mean removing intrinsics (arguably this was already not the case with the tiny/huge). But what we really do here is express the loops explicitly, so maybe we want to postfix this transformations with "2loop" instead, similar to the arrayrange2loop family of transformations, as the goal is the same.

Also, now that the subclasses are quite light, you could add the last one with a similar pattern: PRODUCT (ARRAY, DIM , MASK)

And instead of listing the intrinsics explicitly in the mms_base you could refer to them as Array Reduction Intrinsics (also in the name of the class and filename).

Other than this renaming suggestions, the inline comments are just small changes.

rupertford commented on src/psyclone/psyir/transformations/intrinsics/maxval2code_trans.py:
```diff
@@ -140,69 +138,37 @@ class Maxval2CodeTrans(MMSBaseTrans):
-          if (maxval_var < array(i_0,i_1)) then
-            maxval_var = array(i_0,i_1)
-          end if
+      result = TINY(result)
+      do idx = 1, 10, 1
+        do idx_1 = 1, 10, 1
+          result = MAX(result, array(idx_1,idx))
```
Oops. Updated for maxval, minval and sum transformations.

rupertford commented on src/psyclone/psyir/transformations/intrinsics/maxval2code_trans.py:
```diff
@@ -140,69 +138,37 @@ class Maxval2CodeTrans(MMSBaseTrans):
+    def _loop_body(self, lhs, rhs):
         '''Provide the body of the nested loop that computes the maximum value
-        of an array.
+        of the lhs and rhs
```
Oops. Changed in maxval, minval and sum transformations.

rupertford commented on src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py:
```diff
@@ -57,6 +56,7 @@ class MMSBaseTrans(Transformation, ABC):
 
     '''
     _INTRINSIC_NAME = None
+    _CHILD_INTRINSIC = None
```
Oops.

rupertford commented on src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py:
```diff
@@ -129,31 +129,36 @@ def validate(self, node, options=None):
-                f"it to be a literal or a reference to a known constant "
-                f"value, but found '{dim_ref.debug_string()}' which is "
-                f"{info}.")
+                f"The dimension argument to {self._INTRINSIC_NAME} is not "
+                f"yet supported.")
```
Sorry for being slow, but why and how does it need updating?

sergisiso replied:
the docstring says: "if a valid value for the dimension argument can't be determined." and now it should be "if it has a dimension value"

rupertford replied:
Ah yes, I missed that you were talking about the docstring. Thanks.

rupertford commented on src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py:
```diff
@@ -129,31 +129,36 @@ def validate(self, node, options=None):
-                (type(array_ref) is Reference)):
+        for reference in array_ref.walk(Reference):
+            if (isinstance(reference, ArrayReference) or
+                    type(reference) == Reference and
```
Updated, thanks.

rupertford commented on src/psyclone/psyir/transformations/intrinsics/sum2code_trans.py:
```diff
@@ -141,64 +140,37 @@ class Sum2CodeTrans(MMSBaseTrans):
-        :type array_ref: :py:class:`psyclone.psyir.nodes.ArrayReference`
+    def _loop_body(self, lhs, rhs):
+        '''Provide the body of the nested loop that computes the sum
+        of the lhs and rhs
```
Updated.

rupertford commented on src/psyclone/psyir/transformations/intrinsics/minval2code_trans.py:
```diff
@@ -140,69 +138,37 @@ class Minval2CodeTrans(MMSBaseTrans):
+    def _loop_body(self, lhs, rhs):
         '''Provide the body of the nested loop that computes the minimum value
-        of an array.
+        of the lhs and rhs
```
Updated.

rupertford commented on src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py:
```diff
@@ -203,170 +187,97 @@ def apply(self, node, options=None):
+                reference2arrayrange.apply(reference)
+            except TransformationError:
+                pass
+        # We know there is only one child
+        rhs = rhs_parent.children[0]
```
The transformation invalidates the value of rhs so we keep a copy of the parent before the transformation and reset the value of rhs using the parent after the transformation. I've added comments to explain. Now whether the transformation should do this or not is another matter!

rupertford commented on src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py:
```diff
@@ -203,170 +187,97 @@ def apply(self, node, options=None):
+        # it on the rhs of an argument with one of the arrays within
+        # the expression being added to the lhs of the argument.
+        rhs = expr.copy()
+        if not rhs.parent:
+            # If the the reference is the only thing on the rhs it
+            # will not have a parent which is required by the
+            # replace_with() method. Add a fake parent (+)
+            _ = UnaryOperation.create(
+                UnaryOperation.Operator.PLUS, rhs)
```
Yes thanks. I think this must have been a remnant from the previous implementation.

rupertford commented on src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py:
```diff
@@ -203,170 +187,97 @@ def apply(self, node, options=None):
+                UnaryOperation.Operator.PLUS, rhs)
+
+        # Convert references to arrays to array ranges where appropriate
+        # Add here to avoid circular imports
+        from psyclone.psyir.transformations import Reference2ArrayRangeTrans
```
I'm not sure what you mean about the pylint top-level warning but you are right about the path. When I am more specific I can declare it at the top of the file.

rupertford commented on src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py:
```diff
@@ -129,31 +129,36 @@ def validate(self, node, options=None):
+                f"Error, no ArrayReference's found in the expression "
+                f"'{array_ref.debug_string()}'.")
+
+        if not node.ancestor(Assignment):
```
Well that was a bit embarrassing. You are right that an intrinsic that was part of an expression would lose the expression. I've now fixed that. We don't need #2387 as the result is always a scalar. When we support the dim argument we will need to worry about arrays but I think we can inherit the properties of the result array if we need a temporary so I suspect we won't need #2387 then either.

rupertford replied:
I added the additional code to `test_multi_intrinsics`, rather than `test_multi`. I have now added the additional code to `test_multi` as well.

rupertford commented:
Renamed the files to from *2code to *2loop. Renamed mms* to array_reduction_*. Updated code, tests and documentation appropriately. Added product2loop transformation. Hopefully these changes address the reviewer's general comments.

rupertford commented:
Ready for next review from @sergisiso

sergisiso commented on src/psyclone/psyir/transformations/intrinsics/array_reduction_base_trans.py:
```diff
@@ -0,0 +1,350 @@
+        orig_lhs = node.ancestor(Assignment).lhs.copy()
+        orig_rhs = node.ancestor(Assignment).rhs.copy()
+        array_refs = rhs.walk(ArrayReference)
+        lhs = array_refs[0].copy()
```
Add the comment "We need the lhs of the created expression to be a reference from the expression because the ArrayRange2Loop will use it to get the loop bounds"

rupertford replied:
Added, thanks.

sergisiso reviewed (changes requested):
@rupertford Almost there, see my response to two of the previous conversations and a small new one.

sergisiso reviewed (approved):
@rupertford All comments have been addressed. Formatters, tests, codecov, documentation building is all green. Integration still doesn't use this transformations to check. This is ready to merge.
